torch/distributed/_shard/replicated_tensor.py

```diff
@@ -0,0 +1,118 @@
+    with ShardedTensor/Tensor together to express different types of computation. The
+    inter-op rules defined as (using torch.add as an example op):
+        ReplicatedTensor + ReplicatedTensor = ReplicatedTensor
+        ReplicatedTensor + torch.Tensor = torch.Tensor
+        ReplicatedTensor + ShardedTensor = ShardedTensor
+
+    NOTE: We do not gurantee equal content of ReplicatedTensor across nodes after its
+    construction. Although we defined proper inter-op rules to make sure ReplicatedTensor
+    stays the same, there's no enforcement on it (i.e. if you manually modify content on
+    some ranks, the modified value will not automatically get synced to other nodes). If
+    you wish to manually validate tensors are the same across ranks, use `validate()`.
+
+    """
+    def __new__(cls, data=None):
+        if data is None:
+            data = torch.empty(0)
+        return torch.Tensor._make_subclass(cls, data)
+
+    def __repr__(self):
+        return f"ReplicatedTensor({super(ReplicatedTensor, self).__repr__()})"
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        new_args = []
+        new_kwargs = {}
+
+        # check if args/kwargs have tensor (non-replicated) operands, we have to do
+        # this so that we can detect there's a torch.Tensor operand and thus not
+        # converting results back to ReplicatedTensor
+        has_tensor = False
+        for arg in args:
+            if isinstance(arg, torch.Tensor) and not isinstance(arg, ReplicatedTensor):
+                has_tensor = True
+
+        if kwargs is not None:
+            for k, v in kwargs.items():
+                if isinstance(v, torch.Tensor) and not isinstance(v, ReplicatedTensor):
+                    has_tensor = True
+
+        # do dispatch base on the inter-op rules we defined
+        for arg in args:
+            if isinstance(arg, ShardedTensor):
+                # redispatch to ShardedTensor
+                # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                return arg.__torch_function__(func, types, args, kwargs)
+            else:
+                new_args.append(arg)
+
+
+        if kwargs is not None:
+            for k, v in kwargs.items():
+                if isinstance(v, ShardedTensor):
+                    # redispatch to ShardedTensor
+                    # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                    return v.__torch_function__(func, types, args, kwargs)
+                else:
+                    new_kwargs[k] = v
+
+        # We cann't do super().__torch_function__() as it implicitly convert the result
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
+            if func in get_default_nowrap_functions():
```

Comment:
For learning purpose. Does this often mean the situation when it's field access like `t.grad`?

Reply:
yeah, more related to field access, this was copied from the `tensor.__torch_function__` because we don't want to go into this function as we manage our own type through the rules we defined.